tiebreaker/src/main/scala/TournamentTiebreaker.scala

```diff
@@ -0,0 +1,279 @@
+    val withLastRoundDrawn: Seq[PlayerGames] = opponentGames.map: opp =>
+      playersWithMaxGames
+        .contains(opp.player)
+        .option(
+          opp.copy(games =
+            opp.games.dropRight(1) ++ opp.games.lastOption.map(_.copy(points = Some(Points.Half))).toSeq
+          )
+        )
+        .getOrElse(opp)
+    buchholzCutN(cut, withLastRoundDrawn)
+
+  private def sonnebornBergerCutN(
+      cut: Int,
+      player: PlayerGames,
+      opponentGames: Seq[PlayerGames]
+  ): TieBreakPoints =
+    player.games
+      .map: game =>
+        val oppScore = opponentGames.find(_.player == game.opponent).map(_.games.score)
+        (oppScore, game.points) match
+          case (Some(score), Some(Points.One))  => score.into(TieBreakPoints)
+          case (Some(score), Some(Points.Half)) => score.map(_ / 2f).into(TieBreakPoints)
+          case _                                => TieBreakPoints(0f)
+      .cutSum(cut)
+
+  private def averageRatingOfOpponentsCutN(
+      cut: Int,
+      opponentGames: Seq[PlayerGames]
+  ): TieBreakPoints =
+    opponentGames
+      .collect:
+        case PlayerGames(Player(_, Some(elo)), _, _) => TieBreakPoints(elo.value)
+      .sorted
+      .drop(cut)
+      .average
+      .map(_.round) // Must round up according to FIDE rules
+
+  private def sumOfProgressiveScoresCutN(
+      cut: Int,
+      player: PlayerGames
+  ): TieBreakPoints =
+    player.games.indices
+      .map: i =>
+        player.games.take(i + 1).score.into(TieBreakPoints)
+      .cutSum(cut)
+
+  def tb(tiebreaker: Tiebreaker, me: Player, allPlayers: Seq[PlayerGames]): TieBreakPoints =
+    val allMyGames          = allPlayers.find(_.player == me)
+    val allMyOpponentsGames = allPlayers.filter(_.games.exists(_.opponent == me))
+    allMyGames.fold(TieBreakPoints(0f)): meAndMyGames =>
+      meAndMyGames match
+        case PlayerGames(_, myGames, partialTiebreaks) =>
+          tiebreaker match
+            case NbBlackGames => TieBreakPoints(myGames.filter(_.color == Color.Black).size)
+            case NbWins       => TieBreakPoints(myGames.count(_.points.contains(Points.One)))
+            case NbBlackWins  =>
+              TieBreakPoints(myGames.count(g => g.color == Color.Black && g.points.contains(Points.One)))
+            case SonnebornBerger            => sonnebornBergerCutN(0, meAndMyGames, allMyOpponentsGames)
+            case SonnebornBergerCut1        => sonnebornBergerCutN(1, meAndMyGames, allMyOpponentsGames)
+            case Buchholz                   => buchholzCutN(0, allMyOpponentsGames)
+            case BuchholzCut1               => buchholzCutN(1, allMyOpponentsGames)
+            case BuchholzCut2               => buchholzCutN(2, allMyOpponentsGames)
+            case ForeBuchholz               => foreBuchholzCutN(0, allMyOpponentsGames)
+            case ForeBuchholzCut1           => foreBuchholzCutN(1, allMyOpponentsGames)
```

Comment:
this is O(k*n), so to calculate tiebreakers for all players it'll be O(k*n2), not sure it's a big deal or not. We could cache `Buchholz` tiebreak for each players tho.